sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs

```diff
@@ -0,0 +1,253 @@
+        ///
+        protected PluggableCheckpointStoreEventProcessor(CheckpointStore checkpointStore,
+                                                         int eventBatchMaximumCount,
+                                                         string consumerGroup,
+                                                         string fullyQualifiedNamespace,
+                                                         string eventHubName,
+                                                         TokenCredential credential,
+                                                         EventProcessorOptions options = default) : base(eventBatchMaximumCount, consumerGroup, fullyQualifiedNamespace, eventHubName, credential, options)
+        {
+            Argument.AssertNotNull(checkpointStore, nameof(checkpointStore));
+            _checkpointStore = checkpointStore;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="PluggableCheckpointStoreEventProcessor{TPartition}"/> class.
+        /// </summary>
+        ///
+        protected PluggableCheckpointStoreEventProcessor()
+        {
+        }
+
+        /// <summary>
+        ///   Returns a checkpoint for the Event Hub, consumer group, and identifier of the partition associated with the
+        ///   event processor instance, so that processing for a given partition can be properly initialized.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the partition for which to retrieve the checkpoint.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the processing.  This is most likely to occur when the processor is shutting down.</param>
+        ///
+        /// <returns>The checkpoint for the processor to take into account when initializing partition.</returns>
+        ///
+        /// <remarks>
+        ///   Should a partition not have a corresponding checkpoint, the <see cref="EventProcessorOptions.DefaultStartingPosition" /> will
+        ///   be used to initialize the partition for processing.
+        ///
+        ///   In the event that a custom starting point is desired for a single partition, or each partition should start at a unique place,
+        ///   it is recommended that this method express that intent by returning checkpoints for those partitions with the desired custom
+        ///   starting location set.
+        /// </remarks>
+        ///
+        protected override async Task<EventProcessorCheckpoint> GetCheckpointAsync(string partitionId,
+                                                                                   CancellationToken cancellationToken) =>
+            await _checkpointStore.GetCheckpointAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup, partitionId, cancellationToken).ConfigureAwait(false);
```

Comment:
Is it important that we `await` the call here or could we avoid the extra state machine and return the Task directly?

Reply:
I don't think that it's important; but doesn't our general guidance call for awaiting everywhere?   Personally, I don't think that seeing `GetCheckpointAsync` in the stacktrace is helpful.